src/backend/mmap.rs

```diff
@@ -0,0 +1,151 @@
+extern crate memmap;
+extern crate failure;
+
+use self::failure::ResultExt;
+
+use super::Backend;
+
+use ::error;
+use ::std::io;
+use ::std::cmp;
+
+#[derive(Debug)]
+struct Mmap {
+    inner: memmap::MmapMut,
+    //End of data
+    pub end: usize,
+    //Mmap total len
+    pub len: usize
+}
+
+impl Mmap {
+    fn new(len: usize) -> io::Result<Self> {
+        let inner = memmap::MmapOptions::new().len(len).map_anon()?;
+
+        Ok(Self {
+            inner,
+            end: 0,
+            len
+        })
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        &self.inner[..self.end]
+    }
+
+    fn as_mut_slice(&mut self) -> &mut [u8] {
+        &mut self.inner[..self.end]
+    }
+
+    //Copies data to mmap and modifies data's end cursor.
+    fn write(&mut self, data: &[u8]) {
+        assert!(data.len() <= self.len);
+        self.end = data.len();
+        self.as_mut_slice().copy_from_slice(data);
+    }
+
+    #[inline]
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner.flush()
+    }
+
+    //Increases mmap size by max(old_size*2, new_size)
+    //Note that it doesn't copy original data
+    fn extend(&mut self, new_size: usize) -> io::Result<()> {
+        let len = cmp::max(self.len + self.len, new_size);
+        //Make sure we don't discard old mmap before creating new one;
+        let new_mmap = Self::new(len)?;
+        *self = new_mmap;
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+/// A backend that uses anonymous mmap
+///
+/// The Backend automatically creates bigger map
+/// on demand using following strategy approach:
+///
+/// - If new data size allows, multiply size by 2.
+/// - Otherwise new data size
+///
+/// Note that mmap is never shrink back.
+///
+/// Use `Backend` methods to read and write into it.
+pub struct MmapStorage {
+    mmap: Mmap
+}
+
+impl MmapStorage {
+    #[inline]
```

Comment:
Consider removing `inline` as it is not needed most of the time. LLVM will optimize this correctly and inline it anyway